cmd/clusterctl/client/tree/discovery.go

```diff
@@ -93,54 +104,131 @@ func Discovery(ctx context.Context, c client.Client, namespace, name string, opt
 
 		if visible {
 			if machineInfra, err := external.Get(ctx, c, &m.Spec.InfrastructureRef, cluster.Namespace); err == nil {
-				tree.Add(m, machineInfra, ObjectMetaName("MachineInfrastructure"), NoEcho(true))
+				tree.Add(m, machineInfra, ObjectMetaName("MachineInfrastructure"), NoEcho(!options.ShowAllResources))
 			}
 
 			if machineBootstrap, err := external.Get(ctx, c, m.Spec.Bootstrap.ConfigRef, cluster.Namespace); err == nil {
-				tree.Add(m, machineBootstrap, ObjectMetaName("BootstrapConfig"), NoEcho(true))
+				tree.Add(m, machineBootstrap, ObjectMetaName("BootstrapConfig"), NoEcho(!options.ShowAllResources))
 			}
 		}
 	}
 
 	controlPlaneMachines := selectControlPlaneMachines(machinesList)
 	for i := range controlPlaneMachines {
 		cp := controlPlaneMachines[i]
-		addMachineFunc(controlPLane, cp)
+		addMachineFunc(controlPlane, cp)
+	}
+
+	machinePoolList, err := getMachinePoolsInCluster(ctx, c, cluster.Namespace, cluster.Name)
+	if err != nil {
+		return nil, err
 	}
 
-	if len(machinesList.Items) == len(controlPlaneMachines) {
+	if len(machinesList.Items) == len(controlPlaneMachines) && len(machinePoolList.Items) == 0 {
+		// TODO: revise this check when machinepoolmachines is implemented
 		return tree, nil
 	}
 
 	workers := VirtualObject(cluster.Namespace, "WorkerGroup", "Workers")
 	tree.Add(cluster, workers)
 
+	err = addMachineDeploymentToObjectTree(ctx, c, cluster, workers, machinesList, tree, options, addMachineFunc)
+	if err != nil {
+		return nil, err
+	}
+
+	addMachinePoolsToObjectTree(ctx, c, cluster, workers, machinePoolList, tree, options)
+
+	// Handles orphan machines.
+	if len(machineMap) < len(machinesList.Items) {
+		other := VirtualObject(cluster.Namespace, "OtherGroup", "Other")
+		tree.Add(workers, other)
+
+		for i := range machinesList.Items {
+			m := &machinesList.Items[i]
+			if _, ok := machineMap[m.Name]; ok {
+				continue
+			}
+			addMachineFunc(other, m)
+		}
+	}
+
+	return tree, nil
+}
+
+func addAddonsToObjectTree(ctx context.Context, c client.Client, cluster *clusterv1.Cluster, tree *ObjectTree) {
+	if resourceSetBindings, err := getResourceSetBindingsInCluster(ctx, c, cluster.Namespace, cluster.Name); err == nil {
+		// Should this be in it's own group?
+		addons := VirtualObject(cluster.Namespace, "AddonGroup", "Addons")
+		tree.Add(cluster, addons)
+		for i, resourceSetBinding := range resourceSetBindings {
+			tree.Add(addons, &resourceSetBindings[i], ObjectMetaName("ClusterResourceSetBinding"))
+			for _, binding := range resourceSetBinding.Spec.Bindings {
+				if resourceSet, err := getResourceSetInCluster(ctx, c, cluster.Namespace, binding.ClusterResourceSetName); err == nil {
```

Comment:
Is it required to read the actual CRS, given that we are not using info contained in it?